andrepereiradasilva/joomla-cms
# Improve ajax js errors and solve ACL problems

Pull Request for Several ACL issues.
#### Summary of Changes

This PR makes several changes in the permissions panel and also replicate js improvements in sendtest mail in order to solves bugs and improve php/js code.
###### Bugs/Improvements

1: Ajax js errors are not catched in permission.js
After this PR uses the same checks as sendtestmail.js that where now moved to core.js

2: Session token is not being checked in permission.js
After this PR the session token is checked.

3: When a "Denied" permission is inherited the calculated permission is "Not Allowed (Locked)"
After this PR the text was changed to "Not Allowed (Inhreited)"

4: Permision js ajax call is returned in html mimetype
After this PR the ajax call is returned in json mimetype.

5: Some errored of permision and sentestmil js ajax calls are returned in HTML
After this PR the all messages are returned in Json.

6: The js messages in sendtestmail and permissions.js are not cleaned on change.
After this PR they are cleaned everytime you change.

7: Permission js uses a custom html message system
After this PR all message use the Joomla API message system.

8: A super user cannot remove super user permission from other groups that his not into.
After this PR this is possible.

9: When a Super permission is changed or a a group with child groups permission is changed the user is not informed that we needs to refresh the page to recalculate permissions.
After this PR the user is notified.

10: A user that is not Super User can change a Super User group permissions.
After this PR this is not possible.

11: A user that is not Super User can change change its own's groups ot it's parent permissions.
After this PR this is not possible.

12: We don't have calulate permissions in com_config public group.
After this PR we have the calculated permission here too.

13: Changing a permission with "Not Allowed (Inhreited)" to "Alloewd" or "Denied" changes it's calculate permission. 
After this PR the calculated permissions always stays "Not Allowed (Inhreited)" in this cases.

14: If you change a "Inherited" permission with a calculated permission of "Not Allowed" to "Allowed" and then back to "Inherited", the calculated status is changed to "Allowed" (it should be "Not Allowed").
This PR solves that.

15: There are some "Conflict" messages when you have a parent group with "Denied" and some child group with "Allowed". This should be only "Not Allowed (Inhreited)"
This PR solves that.

There can be more bugs/improvements that i don't remember...
##### Scenarios
###### Calculated Permissions Scenarios

If only "Inherited" across the group tree:
- "Inherited" -> "Not Allowed" (red).
- "Allowed" -> "Allowed" (green).
- "Denied" -> "Not Allowed" (red).

If any "Denied" across the group tree:
- "Inherited" -> "Not Allowed (Inherited)" (gray).
- "Allowed" -> "Not Allowed (Inherited)" (gray).
- "Denied" -> "Not Allowed (Inherited)" (gray).

If no "Denied" and a "Allowed" across the group tree:
- "Inherited" -> "Allowed" (green).
- "Allowed" -> "Allowed" (green).
- "Denied" -> "Not Allowed" (red).

_Special case_: If group has super User permissions all values in that group and child groups should be "Allowed (Super User)"  (green).
###### Error/Notices Scenarios

Can't change permissions scenarios:
- Session token is not valid.
- Database query/update errors.
- Any error while checking or storing asset.
- Current user does not have permissions to change ACL for the component.
- Current user is Super User and is trying to remove his own Super User permissions.
- Current user is not Super User and is trying to change a Super User group permissions.
- Current user is not Super User and is trying to change its own's groups permissions.
- Current user is not Super User and is trying to change its own's groups parent permissions.

Ask user to reload the page scenarios:
- User changed permissions of a group with child groups.
- User changed group permissions to Super User.
#### Testing Instructions
- Use latest staging.
- Check all "Bugs/Improvements".
- Apply patch
- Go to global configuration
- Clean browser cache.
- Check all "Bugs/Improvements" are solved.
- Check all scenarios described in the "Scenarios".
- Do the tests two times. With your super user and other user with less permissions (easier if you open a session for that user in a private browser window).
